rust-engine/src/backends.rs

```diff
@@ -14,14 +14,107 @@
+    ast::{Item, Metadata, Module, span::Span},
+    printer::{Print, Printer},
+};
+use hax_types::engine_api::File;
+use std::path::PathBuf;
+
+/// A hax backend.
+///
+/// A backend is responsible for turning the hax AST into sources of a target language.
+/// It combines:
+/// - a sequence of AST transformation phases, and
+/// - a printer that generates textual output.
+///
+/// For example, we have F\*, Coq, and Lean backends.
+/// Some are still in the old OCaml engine.
+pub trait Backend {
+    /// The printer type used by this backend.
+    type Printer: Printer;
+
+    /// Construct a new printer instance.
+    ///
+    /// By default this calls `Default::default` on the printer type.
+    fn printer(&self) -> Self::Printer {
+        Self::Printer::default()
+    }
+
+    /// A short name identifying the backend.
+    ///
+    /// By default, this is delegated to the associated printer's [`Printer::NAME`].
+    const NAME: &'static str = Self::Printer::NAME;
+
+    /// The AST phases to apply before printing.
+    ///
+    /// Backends can override this to add transformations.  
+    /// The default is an empty list (no transformations).
+    fn phases(&self) -> Vec<Box<dyn crate::phase::Phase>> {
+        vec![]
+    }
+
+    /// Group a flat list of items into modules.
+    ///
+    /// By default, everything is packed into a single module, since we
+    /// don't yet support proper name rendering (see issue #1599).
+    fn items_to_module(&self, items: Vec<Item>) -> Vec<Module> {
+        let Some(first) = items.first() else {
+            return vec![];
+        };
+        vec![Module {
+            ident: first.ident.clone(),
+            items,
+            meta: Metadata {
+                span: Span::dummy(),
+                attributes: vec![],
+            },
+        }]
+    }
+
+    /// Compute the relative filesystem path where a given module should be written.
+    fn module_path(&self, module: &Module) -> PathBuf;
+}
+
+/// Apply a backend to a collection of AST items, producing output files.
+///
+/// This runs all of the backend's [`Backend::phases`], groups the items into
+/// modules via [`Backend::items_to_module`], and then uses the backend's printer
+/// to generate source files with paths determined by [`Backend::module_path`].
+pub fn apply_backend<B: Backend + 'static>(backend: B, mut items: Vec<Item>) -> Vec<File> {
+    for phase in backend.phases() {
+        phase.apply(&mut items);
+    }
+
+    let modules = backend.items_to_module(items);
+    modules
+        .into_iter()
+        .map(|module: Module| {
+            let path = backend
+                .module_path(&module)
+                .into_os_string()
+                .into_string()
+                .unwrap();
+            let (contents, _) = backend.printer().print(module);
+            File {
+                path,
+                contents,
+                sourcemap: None,
+            }
+        })
+        .collect()
+}
+
 #[allow(unused)]
```

Comment:
Can this really be unused? Looks like it's used.

Reply:
Currently, with only a dummy Rust printer, some things are unused: the `Pretty` trait for instance.
But non-dummy printer commonly needs that trait.

Reply:
Maybe the `allow(unused)` should then go on the bits that are actually unused?

Reply:
I opened https://github.com/cryspen/hax/issues/1615